core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,272 @@
+        }
+
+    var sendErrorLogData: Boolean
+        get() = _sendErrorLogData
+        set(value) {
+            _sendErrorLogData = value
+        }
+    var errorHandler: ((Throwable) -> Unit)?
+        get() = _errorHandler
+        set(value) {
+            _errorHandler = value
+        }
+
+    var host: String
+        get() = _host
+        set(value) {
+            _host = value
+        }
+
+    var sampleRate: Double
+        get() = _sampleRate
+        set(value) {
+            _sampleRate = value
+        }
+
+    var flushTimer: Int
+        get() = _flushTimer
+        set(value) {
+            _flushTimer = value
+        }
+
+    var maxQueueSize: Int
+        get() = _maxQueueSize
+        set(value) {
+            _maxQueueSize = value
+        }
+
+    var maxQueueBytes: Int
+        get() = _maxQueueBytes
+        set(value) {
+            _maxQueueBytes = min(value, MAX_QUEUE_BYTES)
+        }
+
+    var httpClient: HTTPClient
+        get() = _httpClient
+        set(value) {
+            _httpClient = value
+        }
+
+    private val queue = mutableListOf<RemoteMetric>()
+    private var queueBytes = 0
+    private var queueSizeExceeded = false
+    private val seenErrors = mutableMapOf<String, Int>()
+    private var started = false
+
+    init {
+        CoroutineScope(Dispatchers.Default).launch {
+            while (isActive) {
+                if (!enable) return@launch
+                try {
+                    flush()
+                } catch (e: Throwable) {
+                    logError(e)
+                }
+                try {
+                    delay(flushTimer.toLong())
+                } catch (e: CancellationException) {
+                    // last flush before shutdown
+                    flush()
+                }
+
+            }
+        }
+    }
+
+    fun start() {
+        if (started) return
+        // Assume sampleRate is now set and everything in the queue hasn't had it applied
+        if (Math.random() > sampleRate) {
+            resetQueue()
+        } else {
+            queue.forEach {
+                it.value = (it.value / sampleRate).roundToInt()
+            }
+        }
+
+        started = true
+    }
+
+    fun increment(metric: String, tags: Map<String, String>) {
+        if (!enable) return
+        if (!metric.startsWith("analytics_mobile.")) return
+        if (tags.isEmpty()) return
+        if (Math.random() > sampleRate) return
+        if (queue.size >= maxQueueSize) return
+
+        addRemoteMetric(metric, tags, value = (1.0 / sampleRate).roundToInt())
+    }
+
+    fun error(metric:String, tags: Map<String, String>, log: String) {
+        if (!enable) return
+        if (!metric.startsWith("analytics_mobile.")) return
+        if (tags.isEmpty()) return
+        if (queue.size >= maxQueueSize) return
+
+        var filteredTags = tags
+        if (!sendWriteKeyOnError) filteredTags = tags.filterKeys { it != "writekey" }
```

Comment:
We use about the case here? do we need to toLowerCase() the key before checking?

Reply:
Apparently it's .lowercase() now, thanks.